File: problems/5/solution3.ts
// Accepted: 87ms (65.00%), 45.32MB (86.00%)

// sliding window

function isPalindrome(s: string, beg: number, end: number): boolean {
    let left = beg;
    let right = end;

    while (left <= right) {
        if (s[left] !== s[right]) {
            return false;
        }

        left++;
        right--;
    }

    return true;
}

function scanLongestPalindrome(s: string, beg: number, end: number): string {
    const n = s.length;
    let longestBeg = -1;
    let longestEnd = -1;

    // scan palindrome with window, from left to right
    while (end < n) {
        if (!isPalindrome(s, beg, end)) {
            beg++;
            end++;
            continue;
        }

        longestBeg = beg;
        longestEnd = end;

        // move window to right, if end of boundary
        if (beg === 0 || end === n-1) {
            beg++;
            end++;
            continue;
        }

        // expand window
        beg--;
        end++;
    }

    // not found
    if (longestBeg === -1 || longestEnd === -1) {
        return "";
    }

    return s.slice(longestBeg, longestEnd+1);
}

function longestPalindrome(s: string): string {
    const oddLength = scanLongestPalindrome(s, 0, 0);
    const evenLength = scanLongestPalindrome(s, 0, 1);

    return oddLength.length > evenLength.length ? oddLength : evenLength;
};

File: problems/5/solution3.py
# Accepted: 349ms (86.80%), 17.32MB (31.95%)

# sliding window

class Solution:
    def isPalindrome(self, s: str, beg: int, end: int) -> bool:
        left = beg
        right = end

        while left <= right:
            if s[left] != s[right]:
                return False

            left += 1
            right -= 1

        return True

    def scanLongestPalindrome(self, s: str, beg: int, end: int) -> string:
        n = len(s)
        longestBeg = -1
        longestEnd = -1

        # scan palindrome with window, from left to right
        while end < n:
            if not self.isPalindrome(s, beg, end):
                beg += 1
                end += 1
                continue

            longestBeg = beg
            longestEnd = end

            # move window to right, if end of bounary
            if beg == 0 or end == n-1:
                beg += 1
                end += 1
                continue

            # expand window
            beg -= 1
            end += 1

        # not found
        if longestBeg == -1 or longestEnd == -1:
            return ""

        return s[longestBeg:longestEnd+1]

    def longestPalindrome(self, s: str) -> str:
        oddLength = self.scanLongestPalindrome(s, 0, 0)
        evenLength = self.scanLongestPalindrome(s, 0, 1)

        return oddLength if len(oddLength) > len(evenLength) else evenLength
